### PR TITLE
Refactor/watch prev kv

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
based on #268 
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Handle prev_kv in watch
* what changes does this pull request make?
Reduce unnecessary io operations about prev_kv, and read it before notify when necessary
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xline/src/server/watch_server.rs [hunk 1]
```diff
@@ -406,12 +433,95 @@ mod test {
...
+            Arc::clone(&kv_watcher),
+            res_tx,
+            req_stream,
+            Arc::clone(&header_gen),
+        ));
+
+        for _ in 0..4 {
+            let watch_res = res_rx.recv().await.unwrap().unwrap();
+            if watch_res.events.is_empty() {
+                // create response
```
Comment (Phoenix500526) [inline xline/src/server/watch_server.rs:495]: create response?

Comment (themanforfree) [reply xline/src/server/watch_server.rs]: Fixed, it should be `WatchCreateResponse`

File: xline/src/storage/kv_store.rs [hunk 2]
```diff
@@ -728,21 +723,39 @@ where
...
+            &self.lease_collection,
+            &req.key,
+            &req.range_end,
+            revision,
+            sub_revision,
+        )
+    }
+
+    /// Delete keys from index and detach them in lease collection, return all the write operations and events
+    pub(crate) fn delete_keys<'a>(
```
Comment (Phoenix500526) [inline xline/src/storage/kv_store.rs:739]: `pub(super)` is better.

File: xline/src/storage/kv_store.rs [hunk 3]
```diff
@@ -376,6 +376,13 @@ where
         Ok((kvs, total))
     }
 
+    /// Get previous `KeyValue` of a `KeyValue`
+    pub(crate) fn get_prev_kv(&self, kv: &KeyValue) -> Option<KeyValue> {
```
Comment (Phoenix500526) [inline xline/src/storage/kv_store.rs:380]: `get_prev_kv` is invoked in `KvWatcher`. Both `kv_store` and `kvwatcher` are belonged to `storage`. So, I think `pub(super)` is enough

File: xline/src/storage/kvwatcher.rs [hunk 4]
```diff
@@ -564,20 +576,16 @@ mod test {
         let handle = tokio::spawn({
             let store = Arc::clone(&store);
             async move {
-                let mut revision = 2;
+                // let mut revision = 2;
```
Comment (Phoenix500526) [inline xline/src/storage/kvwatcher.rs:579]: remove this line of code if we don't need it.

File: xline/src/storage/kvwatcher.rs [hunk 5]
```diff
@@ -288,6 +292,9 @@ pub(crate) trait KvWatcherOps {
 
     /// Cancel a watch from KV store
     fn cancel(&self, id: WatchId);
+
+    /// Get Prev `KeyValue` of a `KeyValue`
+    fn get_prev_kv(&self, kv: &KeyValue) -> Option<KeyValue>;
```
Comment (Phoenix500526) [inline xline/src/storage/kvwatcher.rs:297]: Shall we make a better name? This method's name is a little bit confused with `KvStore::get_prev_key`

Comment (themanforfree) [reply xline/src/storage/kvwatcher.rs:297]: we don't have a method called `KvStore::get_prev_key`, we only have `KvStore::get_prev_kv`, and this method calls `KvStore::get_prev_key`, they did the same thing

File: xline/src/storage/kvwatcher.rs (hunk 4 above)
Comment (themanforfree) [reply xline/src/storage/kvwatcher.rs:579]: Removed

File: xline/src/storage/kv_store.rs (hunk 3 above)
Comment (themanforfree) [reply xline/src/storage/kv_store.rs:380]: Fixed

File: xline/src/storage/kv_store.rs (hunk 2 above)
Comment (themanforfree) [reply xline/src/storage/kv_store.rs:739]: Fixed

**Review by Phoenix500526 — DISMISSED** [review]

File: xline/src/rpc/mod.rs [hunk 6]
```diff
@@ -475,6 +475,16 @@ impl RequestWithToken {
     }
 }
 
+impl Event {
+    pub(crate) fn is_create(&self) -> bool {
+        let kv = self
+            .kv
+            .as_ref()
+            .unwrap_or_else(|| panic!("kv must be Some"));
+        self.r#type() == EventType::Put && kv.create_revision == kv.mod_revision
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/rpc/mod.rs:484]: Fixed

File: xline/src/server/watch_server.rs [hunk 7]
```diff
@@ -123,6 +123,8 @@ where
     stop_notify: Arc<Event>,
     /// Header Generator
     header_gen: Arc<HeaderGenerator>,
+    /// Prev KV status
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/watch_server.rs:126]: Fixed

**Review by Phoenix500526 — APPROVED** [review]

**Review by rogercloud — APPROVED** [review]

**Pull request merged**